rfc/0003-indexmanagement.md

```diff
@@ -0,0 +1,201 @@
+
+# Watching Indexes
+>**Note**: this part of the API could be considered optional.
+
+Since the process of building a deferred index is asynchronous, the server immediately returns control to the client. It would be useful to have an API to poll the list of indexes until one (or several) indexes are seen as "online". In the blocking API, this would allow to make sure that an index is ready.
+
+Signature:
+
+```java
+List<IndexInfo> watchIndexes(List<String> watchList, boolean watchPrimary, long watchTimeout, TimeUnit watchTimeUnit)
+```
+
+Remarks:
+The implementation would be based on regularly polling the index view (using `listIndexes()` for instance) until a specific index (or a list of indexes) is seen as `online`.
+
+`boolean watchPrimary`, if set to `true`, would have the effect of adding the default primary name **`#primary`** to the watchList. It is recommended to have that name as a constant somewhere and reference it in the method inline documentation (for instance in Java this is in `Index.PRIMARY_NAME` in the dsl).
+
+The `watchTimeout` and `watchTimeUnit` are a way to express a maximum **duration** for the poll. This can be replaced by a language-idiomatic way of representing a duration in each SDK.
+
+An index that doesn't exist is ignored. `watchIndexes` returns the `List<IndexInfo>` of indexes that
+
+ 1. were already online or...
+ 2. became online during the watch timeframe.
+
+Thus it can be empty if indexes were all "pending"/"deferred"/"building" and failed to be built during the watch timeframe, or if no index in the provided list exists.
+
+If this is easy to implement, `watchIndexes` should rely on a polling with a linearly augmenting delay (50ms to 1000ms by steps of 500ms is deemed a good retry pattern). In such a case, each poll cycle may rely on a `IndexesNotReadyException` internally to signal that current cycle didn't observe all requested indexes.
+
+For SDKs with an idiomatic asynchronous API (eg. Java with RxJava), an asynchronous version of the polling should be implemented, allowing to avoid blocking the rest of the code if an index takes a long time to become online.
+
+For SDKs with configurable logging semantics, a mean of toggling a DEBUG log of the different polling attempts may be offered.
+
+# Language Quick Example and Specifics
+## Java
+
+```java
+//create a primary index, if there is none
+bucketManager.createPrimaryIndex(true, false);
+
+//make sure "byDescAndToto" will be recreated from scratch by dropping it if it exists
+bucketManager.dropIndex("byDescAndToto", true);
+
+//defer creation of the secondary index on field toto and desc.
+//note that DESC is a keyword, but the field is escaped when provided as a String.
+bucketManager.createIndex("byDescAndToto", false, true, x("toto"), "desc");
+
+//since one of the indexes was built with defer, trigger the build and wait for it to finish in maximum 1 minute
+bucketManager.watchIndexes(bucketManager.buildDeferredIndexes(), 1, TimeUnit.MINUTES);
+
```

Comment:
I'll edit one last time to clarify that `watchIndexes` is indeed optional and to be implemented in the manner that seems the most idiomatic to each particular SDK. In Java, since we have clearly defined sync/async APIs which mirror each other, this translates to a `List<IndexInfo> watchIndexes` method in the blocking API mirrored by a `Observable<IndexInfo> watchIndexes` in the async API.